spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestDeleteReachableFilesProcedure.java

```diff
@@ -0,0 +1,117 @@
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import static org.apache.iceberg.TableProperties.GC_ENABLED;
+
+public class TestDeleteReachableFilesProcedure extends SparkExtensionsTestBase {
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  public TestDeleteReachableFilesProcedure(String catalogName, String implementation, Map<String, String> config) {
+    super(catalogName, implementation, config);
+  }
+
+  @Test
+  public void testDeleteReachableFilesInEmptyTable() throws IOException {
+    if (!createTable()) {
+      return;
+    }
+    Table table = catalog.loadTable(tableIdent);
+    String metadataFileLocation = metadataFileLocation(table);
+
+    catalog.dropTable(tableIdent, false);
+
+    List<Object[]> output = sql(
+        "CALL %s.system.delete_reachable_files('%s')",
+        catalogName, metadataFileLocation);
+    assertEquals("Should delete only other files (metadata json and version_hint file)",
+        ImmutableList.of(row(0L, 0L, 0L, 2L)), output);
+  }
+
+  @Test
+  public void testDeleteReachableFilesInDataFolder() throws IOException {
+    if (!createTable()) {
+      return;
+    }
+
+    sql("INSERT INTO TABLE %s VALUES (1, 'a')", tableName);
+    sql("INSERT INTO TABLE %s VALUES (2, 'b')", tableName);
+
+    Table table = catalog.loadTable(tableIdent);
+    String metadataFileLocation = metadataFileLocation(table);
+
+    catalog.dropTable(tableIdent, false);
+
+    List<Object[]> output = sql(
+        "CALL %s.system.delete_reachable_files('%s')",
+        catalogName, metadataFileLocation);
+    assertEquals("Should delete expected number of files",
+        ImmutableList.of(row(2L, 2L, 2L, 4L)), output);
+  }
+
+  @Test
+  public void testDeleteReachableFilesGCDisabled() throws IOException {
+    if (!createTable()) {
+      return;
+    }
+    sql("ALTER TABLE %s SET TBLPROPERTIES ('%s' 'false')", tableName, GC_ENABLED);
+
+    Table table = catalog.loadTable(tableIdent);
+    String metadataFileLocation = metadataFileLocation(table);
+
+    catalog.dropTable(tableIdent, false);
+    AssertHelpers.assertThrows("Should reject call",
+        ValidationException.class, "Cannot remove files: GC is disabled",
+        () -> sql("CALL %s.system.delete_reachable_files('%s')", catalogName, metadataFileLocation));
+  }
+
+  private String metadataFileLocation(Table tbl) {
+    return ((HasTableOperations) tbl).operations().current().metadataFileLocation();
+  }
+
+  private boolean createTable() throws IOException {
+    if (catalogName.equals("testhadoop") || catalogName.equals("testhive")) {
+      // This procedure cannot work for hadoop catalog,
+      // as after drop table metadata file will be deleted (even with purge=false)
+      // For hive catalog, drop table with purge = true is not cleaning the table in metastore.
```

Comment:
Is it an issue ?

Reply:
An example of setting the `isHadoopCatalog` variable in that same file: https://github.com/apache/iceberg/blob/f5a753791f4dc6aca78569a14f731feda9edf462/spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/sql/TestNamespaceSQL.java#L43-L49

Notice that it is the same call, but it's much cleaner to look at in the test when you use `Assume.assumeFalse(isHadoopCatalog)`.

Reply:
But more importantly (and sorry for the long winded comments, but I figured since you're asking these are useful things we use a lot in the repo):

I'd suggest getting the temp folder to be in the correct state in either an `@Before` function or an `@After` function so this can be run with as many catalogs as are supported by it. It's usually just `HadoopCatalog` or `spark_catalog` that I see the assumption on.

So the HiveCatalog assumption gives me a bit of pause and makes me wonder if we can't fix it temporarily within the test and then fix the behavior fully in a later patch (please open a ticket if there is an incorrect behavior taking place).

Also, have you merged in latest master? Somebody submitted a fix for properly purging data files so that might help. But this seems to be about NOT wanting to purge and having purge happen anyways (if my understanding is correct), so that seems like a new issue that needs to be looked into.